### PR TITLE
Add compiler support for PostgreSQL "NOT VALID" constraints.

### LINKED ISSUE #7600: Implement compiler support for NOT VALID PostgreSQL constraints
### Describe the use case

Currently, it is not possible to declare `ForeignKeyContraint`s and `CheckConstraint`s as `NOT VALID`.

`NOT VALID` constraints only work with PostgreSQL 9.1+ and only exist for `FOREIGN KEY` and `CHECK` constraints. Setting a constraint as `NOT VALID` will prevent PostgreSQL to check the constraint when creating it.

See https://www.postgresql.org/docs/current/sql-altertable.html

### Databases / Backends / Drivers targeted

- PostgreSQL 9.1+

### Example Use

```python
Table(
    'students',
    meta,
    Column('id', Integer, primary_key=True),
    Column("head_teacher_id", Integer)
    Column('lastname', String),
    CheckConstraint('lastname IS NOT NULL', postgresql_not_valid=True),
    ForeignKeyConstraint(["head_teacher_id"], ["teachers.id"], postgresql_not_valid=True),
)
```

### Additional context

- SO question: https://stackoverflow.com/questions/70803156/how-to-declare-not-valid-postgresql-constraints-with-sqlalchemy-and-alembic
- #4825 implemented introspection support for `NOT VALID`, but no compiler support
- This option doesn't make much sense in `CREATE TABLE`, although PostgreSQL seems to be fine with it. However, having this supported in SQLAlchemy would be a preliminary step to having this working in Alembic, which would definitely be useful.

### PR DESCRIPTION
**Pull request by gilbsgilbs**
### Description

PostgreSQL supports declaring CHECK and FOREIGN KEY constraints as "NOT
VALID" which prevents the constraint from being validated on already
inserted rows, but still validates against any newly inserted or updated
rows.

This commit adds compiler support for this dialect option on
`CheckConstraint` and `ForeignKeyConstraint` classes. Although this
option doesn't make much sense on table creation, it doesn't seem to
cause any issue with Postgres and it is a preliminary step to have this
supported in Alembic during `ALTER TABLE`. Note that other constraints
types do not support this option.

See https://www.postgresql.org/docs/current/sql-altertable.html

See also #4825 which added supported for "NOT VALID" constraints
introsecpection.

Fixes #7600

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

File: lib/sqlalchemy/dialects/postgresql/base.py [hunk 1]
```diff
@@ -2580,6 +2596,14 @@ def get_column_specification(self, column, **kwargs):
             colspec += " NULL"
         return colspec
 
+    def _define_constraint_validity(self, constraint):
+        if self.dialect._supports_not_valid_constraints:
+            not_valid = constraint.dialect_options["postgresql"]["not_valid"]
+            if not_valid:
+                return " NOT VALID"
+
+        return ""
```
Comment (zzzeek) [inline lib/sqlalchemy/dialects/postgresql/base.py:2605]: it looks like NOT VALID doesn't actually change the functionality of the constraint so I guess it is safe to silently ignore for older versions; but in any case version 9.1 is super old at this point so it doesn't matter much anyway.

**Review by zzzeek — COMMENTED** [review]
not bad, looks pretty straightforward, thanks

File: lib/sqlalchemy/dialects/postgresql/base.py (hunk 1 above)
Comment (zzzeek) [reply lib/sqlalchemy/dialects/postgresql/base.py:2605]: our "normal support" is based at 9.6: https://docs.sqlalchemy.org/en/14/dialects/postgresql.html  so I think just render the directive if it's asked for should be fine here

Comment (gilbsgilbs) [reply lib/sqlalchemy/dialects/postgresql/base.py]: Fair enough. c2ed3e9b1c2e8510e139f40665bfcffcbf12e23d

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision c2ed3e9b1c2e8510e139f40665bfcffcbf12e23d of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change c2ed3e9b1c2e8510e139f40665bfcffcbf12e23d: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3557

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

needs a 1.4 backport + a changelog note



View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3557

**Comment by gilbsgilbs** [comment]
@zzzeek I just added a changelog note for 1.4. Regarding the backport though, I'm not sure what you're expecting from me. Do you want me to open another PR or change my PR target branch to `rel_1_4`?

**Comment by sqla-tester** [comment]
**mike bayer** (zzzeek) wrote:

> Patch Set 1:
> 
> gilbsgilbs@github wrote:
> 
> @zzzeek I just added a changelog note for 1.4. Regarding the backport though, I'm not sure what you're expecting from me. Do you want me to open another PR or change my PR target branch to `rel_1_4`?

we can do it on this end, my note was out there for whomever wanted to finish this up, often @caselit or myself will go through and finish up PRs that need backports.  it's alittle tedious to do with pull requests so I wouldn't worry about it, thanks for writing the change note!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3557

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 78eecd55fd9fad07030d963f5fd6713c4af60e80 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 78eecd55fd9fad07030d963f5fd6713c4af60e80 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3557

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3557 has been **merged**. Congratulations! :)

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3568 has been **merged**. Congratulations! :)

**Comment by gilbsgilbs** [comment]
Thanks a lot @zzzeek .

**Pull request closed without merging**